server/services/v1/query_runner.go

```diff
@@ -962,13 +978,78 @@ func (runner *SearchQueryRunner) getSortOrdering(coll *schema.DefaultCollection)
 	return ordering, nil
 }
 
+type PublishQueryRunner struct {
+	*BaseQueryRunner
+
+	req *api.PublishRequest
+}
+
+func (runner *PublishQueryRunner) Run(ctx context.Context, tx transaction.Tx, tenant *metadata.Tenant) (*Response, context.Context, error) {
+	db, err := runner.getDatabase(ctx, tx, tenant, runner.req.GetDb())
+	if err != nil {
+		return nil, ctx, err
+	}
+
+	ctx = runner.cdcMgr.WrapContext(ctx, db.Name())
+
+	coll, err := runner.getCollection(db, runner.req.GetCollection())
+	if err != nil {
+		return nil, ctx, err
+	}
+
+	ts, allKeys, err := runner.publish(ctx, tx, tenant, db, coll, runner.req.GetMessages())
+	if err != nil {
+		return nil, ctx, err
+	}
+	return &Response{
+		createdAt: ts,
+		allKeys:   allKeys,
+		status:    PublishedStatus,
+	}, ctx, nil
+}
+
+func (runner *PublishQueryRunner) publish(ctx context.Context, tx transaction.Tx, tenant *metadata.Tenant, db *metadata.Database,
+	coll *schema.DefaultCollection, messages [][]byte) (*internal.Timestamp, [][]byte, error) {
+	var err error
+	var ts = internal.NewTimestamp()
+	var allKeys [][]byte
+	for _, doc := range messages {
+		err = runner.validate(coll, doc)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		table, err := runner.encoder.EncodeTableName(tenant.GetNamespace(), db, coll)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		partition := rand.Intn(partitions)
```

Comment:
Not for now, but this can be moved in some common sharding package/class and have the following:

- GetAnyPartition(totalPartition int)
- GetPartition(key string, totalPartition int)